### PR TITLE
Empty prompt effectively ends with newline

### PR DESCRIPTION
**Pull request by takluyver**
Closes gh-10138

### CODE REVIEW

**Comment by kwankyu** [comment]
Thank you!

File: IPython/terminal/prompts.py [hunk 1]
```diff
@@ -62,11 +62,15 @@ class RichPromptDisplayHook(DisplayHook):
...
-        self.prompt_end_newline = False
+        # If we're not displaying a prompt, it effectively ends with a newline,
+        # because the output will be left-aligned.
+        self.prompt_end_newline = True
+
         if self.do_full_cache:
             tokens = self.shell.prompts.out_prompt_tokens()
-            if tokens and tokens[-1][1].endswith('\n'):
-                self.prompt_end_newline = True
+            if tokens and not tokens[-1][1].endswith('\n'):
```
Comment (kwankyu) [inline IPython/terminal/prompts.py]: The test should be
 `tokens and len(tokens[-1][1]) > 0 and not tokens[-1][1].endswith('\n')`

**Comment by kwankyu** [comment]
Unfortunately, the patch does not fix the issue yet, as an empty prompt still passes the "not ending with newline" test. See the comment attached to the code.

**Comment by kwankyu** [comment]
For your informtion, Sage uses
```
    def out_prompt_tokens(self):
        return [
            (Token.OutPrompt, ''),
        ]   
```

**Comment by kwankyu** [comment]
If you think that a token string should not be empty and Sage should use `[]` for empty output prompt, then your original code will work.

**Comment by takluyver** [comment]
It's not an error for a token string to be empty, but as far as I know there's no need: Sage could simply use an empty list.

Your version might still do the wrong thing if there's an empty final token preceded by one or more non-empty tokens. I've pushed a different fix.

**Comment by kwankyu** [comment]
As you said, empty token string would be of no use, possibly except in the case when the whole prompt is an empty string just as in Sage. So I think we can ignore the impractical case that non-empty tokens followed by an empty final token.

Your fix will work, but it seems to me a bit overdone as the tokens may be rendered twice, once for the test and once for `self.shell.pt_cli.print_tokens`(thought I don't know what this is exactly :-)

**Comment by takluyver** [comment]
'Rendering' them for the check is just concatenating strings together, so I'm not too worried about the performance.

**Comment by kwankyu** [comment]
Ok. I confirm that the patch works well. 

Do I have anything to do for this pull request to be merged and closed, like giving a positive review?

**Comment by takluyver** [comment]
Thanks! There's nothing specific you need to do. We don't normally merge our own PRs, so I'll give other people a few days to merge it. If nobody does and there are no objections, I'll merge it myself.

**Review by willingc — COMMENTED** [review]
@takluyver LGTM

**Pull request merged**